### PR TITLE
[OSPP] Add MatMul and Conv2d optimization pass.

### PR DESCRIPTION
**Pull request by WLFJ**
## MatMul optimization detail
I added a pass working like `--convert-linalg-to-affine-loops`, but the result is more complex.
In pass we can specify vector size via vec-size (How many element compute simultaneously), kernel-m means the number of rows in kernel, and kernel-n for column number.

command:
```bash
buddy-opt --matmul-optimize="vec-size=16 kernel-m=2 kernel-n=4"
```
input:
```mlir
module{
    func.func @gemm(%a : memref<?x?xf32>, %b : memref<?x?xf32>, %c : memref<?x?xf32>) {
      linalg.matmul 
        ins(%a, %b: memref<?x?xf32>, memref<?x?xf32>)
       outs(%c:memref<?x?xf32>)
      return
    }
}
```
output:
```mlir
#map0 = affine_map<()[s0] -> (s0 - 63)>
#map1 = affine_map<(d0) -> (d0)>
#map2 = affine_map<(d0, d1)[s0, s1] -> (d0 * s1 + s0 + d1)>
#map3 = affine_map<(d0)[s0] -> (d0 + 1, s0 - 1)>
#map4 = affine_map<(d0)[s0] -> (d0, s0 - 1)>
#map5 = affine_map<(d0, d1) -> (0)>
#map6 = affine_map<(d0) -> (d0 + 16)>
#map7 = affine_map<(d0) -> (d0 + 32)>
#map8 = affine_map<(d0) -> (d0 + 48)>
module {
  func.func @gemm(%arg0: memref<?x?xf32>, %arg1: memref<?x?xf32>, %arg2: memref<?x?xf32>) {
    %c0 = arith.constant 0 : index
    %c1 = arith.constant 1 : index
    %c0_0 = arith.constant 0 : index
    %0 = memref.dim %arg0, %c0_0 : memref<?x?xf32>
    %c1_1 = arith.constant 1 : index
    %1 = memref.dim %arg1, %c1_1 : memref<?x?xf32>
    %c1_2 = arith.constant 1 : index
    %2 = memref.dim %arg0, %c1_2 : memref<?x?xf32>
    %3 = affine.apply #map0()[%1]
    affine.for %arg3 = #map1(%c0) to #map1(%3) step 64 {
      affine.for %arg4 = #map1(%c0) to #map1(%0) step 2 {
        %4 = memref.subview %arg0[%arg4, %c0] [%c1, %2] [%c1, %c1] : memref<?x?xf32> to memref<?x?xf32, #map2>
        %5 = affine.min #map3(%arg4)[%0]
        %6 = memref.subview %arg0[%5, %c0] [%c1, %2] [%c1, %c1] : memref<?x?xf32> to memref<?x?xf32, #map2>
        %7 = affine.min #map4(%arg4)[%0]
        %8 = memref.subview %arg2[%7, %c0] [%c1, %1] [%c1, %c1] : memref<?x?xf32> to memref<?x?xf32, #map2>
        %9 = affine.min #map3(%arg4)[%0]
        %10 = memref.subview %arg2[%9, %c0] [%c1, %1] [%c1, %c1] : memref<?x?xf32> to memref<?x?xf32, #map2>
        affine.for %arg5 = #map1(%c0) to #map1(%2) {
          %cst = arith.constant 0.000000e+00 : f32
          %11 = vector.transfer_read %4[%c0, %arg5], %cst {permutation_map = #map5} : memref<?x?xf32, #map2>, vector<16xf32>                                                                                                               
          %cst_3 = arith.constant 0.000000e+00 : f32
          %12 = vector.transfer_read %6[%c0, %arg5], %cst_3 {permutation_map = #map5} : memref<?x?xf32, #map2>, vector<16xf32>                                                                                                             
          %13 = affine.apply #map1(%arg3)
          %cst_4 = arith.constant 0.000000e+00 : f32
          %14 = vector.transfer_read %8[%c0, %13], %cst_4 : memref<?x?xf32, #map2>, vector<16xf32>
          %15 = affine.apply #map6(%arg3)
          %cst_5 = arith.constant 0.000000e+00 : f32
          %16 = vector.transfer_read %8[%c0, %15], %cst_5 : memref<?x?xf32, #map2>, vector<16xf32>
          %17 = affine.apply #map7(%arg3)
          %cst_6 = arith.constant 0.000000e+00 : f32
          %18 = vector.transfer_read %8[%c0, %17], %cst_6 : memref<?x?xf32, #map2>, vector<16xf32>
          %19 = affine.apply #map8(%arg3)
          %cst_7 = arith.constant 0.000000e+00 : f32
          %20 = vector.transfer_read %8[%c0, %19], %cst_7 : memref<?x?xf32, #map2>, vector<16xf32>
          %21 = affine.apply #map1(%arg3)
          %cst_8 = arith.constant 0.000000e+00 : f32                             
          %22 = vector.transfer_read %10[%c0, %21], %cst_8 : memref<?x?xf32, #map2>, vector<16xf32>
          %23 = affine.apply #map6(%arg3)                                                                                                                                                                                                  
          %cst_9 = arith.constant 0.000000e+00 : f32
          %24 = vector.transfer_read %10[%c0, %23], %cst_9 : memref<?x?xf32, #map2>, vector<16xf32>                                                                                                                                        
          %25 = affine.apply #map7(%arg3)                                                        
          %cst_10 = arith.constant 0.000000e+00 : f32                                                                                                                                                                                      
          %26 = vector.transfer_read %10[%c0, %25], %cst_10 : memref<?x?xf32, #map2>, vector<16xf32>
          %27 = affine.apply #map8(%arg3)
          %cst_11 = arith.constant 0.000000e+00 : f32      
          %28 = vector.transfer_read %10[%c0, %27], %cst_11 : memref<?x?xf32, #map2>, vector<16xf32>
          %cst_12 = arith.constant 0.000000e+00 : f32
          %29 = vector.transfer_read %arg1[%arg5, %arg3], %cst_12 : memref<?x?xf32>, vector<16xf32>
          %30 = affine.apply #map6(%arg3)
          %cst_13 = arith.constant 0.000000e+00 : f32
          %31 = vector.transfer_read %arg1[%arg5, %30], %cst_13 : memref<?x?xf32>, vector<16xf32>
          %32 = affine.apply #map7(%arg3)
          %cst_14 = arith.constant 0.000000e+00 : f32                                      
          %33 = vector.transfer_read %arg1[%arg5, %32], %cst_14 : memref<?x?xf32>, vector<16xf32>
          %34 = affine.apply #map8(%arg3)
          %cst_15 = arith.constant 0.000000e+00 : f32
          %35 = vector.transfer_read %arg1[%arg5, %34], %cst_15 : memref<?x?xf32>, vector<16xf32>
          %36 = vector.fma %11, %29, %14 : vector<16xf32>
          %37 = vector.fma %11, %31, %16 : vector<16xf32>
          %38 = vector.fma %11, %33, %18 : vector<16xf32>
          %39 = vector.fma %11, %35, %20 : vector<16xf32>
          %40 = vector.fma %12, %29, %22 : vector<16xf32>
          %41 = vector.fma %12, %31, %24 : vector<16xf32>
          %42 = vector.fma %12, %33, %26 : vector<16xf32>
          %43 = vector.fma %12, %35, %28 : vector<16xf32>                                                     
          %44 = affine.apply #map1(%arg3)
          vector.transfer_write %36, %8[%c0, %44] : vector<16xf32>, memref<?x?xf32, #map2>                 
          %45 = affine.apply #map6(%arg3)
          vector.transfer_write %37, %8[%c0, %45] : vector<16xf32>, memref<?x?xf32, #map2>                 
          %46 = affine.apply #map7(%arg3)
          vector.transfer_write %38, %8[%c0, %46] : vector<16xf32>, memref<?x?xf32, #map2>                  
          %47 = affine.apply #map8(%arg3)           
          vector.transfer_write %39, %8[%c0, %47] : vector<16xf32>, memref<?x?xf32, #map2>
          %48 = affine.apply #map1(%arg3)                                                                                                                                                                                                  
          vector.transfer_write %40, %10[%c0, %48] : vector<16xf32>, memref<?x?xf32, #map2>
          %49 = affine.apply #map6(%arg3)                                                                                                                                                                                                  
          vector.transfer_write %41, %10[%c0, %49] : vector<16xf32>, memref<?x?xf32, #map2>
          %50 = affine.apply #map7(%arg3)           
          vector.transfer_write %42, %10[%c0, %50] : vector<16xf32>, memref<?x?xf32, #map2>       
          %51 = affine.apply #map8(%arg3)
          vector.transfer_write %43, %10[%c0, %51] : vector<16xf32>, memref<?x?xf32, #map2>
        }                                                                                         
      }                                  
    }                                               
    return                                                                                        
  }                                      
}
```                                                               
### Tile Issue
There's lot of min op to ensure the process will not go out of kernel. Redundant calculation not effects the result.

## Conv2d optimization detail

command:
```bash
buddy-opt --conv-optimize="kernel-m=2 kernel-n=3 vec-size=16" Conv2DNchwFchw.mlir
```

input:
```mlir
// generate from iree processed alexnet or lenet mlir file
func.func @conv2d(%input: memref<?x?x?x?xf32>, %filter: memref<?x?x?x?xf32>, %output: memref<?x?x?x?xf32>) {
    linalg.conv_2d_nchw_fchw 
    {dilations = dense<1> : tensor<2xi64>, strides = dense<1> : tensor<2xi64>} 
    ins(%input, %filter : memref<?x?x?x?xf32>, memref<?x?x?x?xf32>) 
    outs(%output : memref<?x?x?x?xf32>)
    return
}
```

output:
```mlir
#map0 = affine_map<(d0) -> (d0)>
#map1 = affine_map<(d0) -> ((d0 ceildiv 2) * 2)>
#map2 = affine_map<(d0, d1) -> (d0 + d1)>
#map3 = affine_map<(d0, d1) -> (d0 + d1 + 16)>
#map4 = affine_map<(d0) -> (d0 + 16)>
#map5 = affine_map<(d0, d1) -> (d0 + d1 + 32)>
#map6 = affine_map<(d0) -> (d0 + 32)>
#map7 = affine_map<(d0, d1) -> (d0 + d1 + 1)>
#map8 = affine_map<(d0) -> (d0 + 1)>
#set = affine_set<(d0)[s0] : (-d0 + s0 - 1 >= 0)>
module {
  func.func @conv2d(%arg0: memref<?x?x?x?xf32>, %arg1: memref<?x?x?x?xf32>, %arg2: memref<?x?x?x?xf32>) {
    %c0 = arith.constant 0 : index
    %cst = arith.constant 0.000000e+00 : f32
    %c0_0 = arith.constant 0 : index
    %0 = memref.dim %arg0, %c0_0 : memref<?x?x?x?xf32>
    %c1 = arith.constant 1 : index
    %1 = memref.dim %arg0, %c1 : memref<?x?x?x?xf32>
    %c2 = arith.constant 2 : index
    %2 = memref.dim %arg0, %c2 : memref<?x?x?x?xf32>
    %c3 = arith.constant 3 : index
    %3 = memref.dim %arg0, %c3 : memref<?x?x?x?xf32>
    %c1_1 = arith.constant 1 : index
    %4 = memref.dim %arg0, %c1_1 : memref<?x?x?x?xf32>
    %c2_2 = arith.constant 2 : index
    %5 = memref.dim %arg0, %c2_2 : memref<?x?x?x?xf32>
    %c3_3 = arith.constant 3 : index
    %6 = memref.dim %arg0, %c3_3 : memref<?x?x?x?xf32>
    %7 = memref.alloc() : memref<1xvector<16xf32>>
    affine.for %arg3 = #map0(%c0) to #map0(%0) {
      affine.for %arg4 = #map0(%c0) to #map0(%1) {
        affine.for %arg5 = #map0(%c0) to #map0(%3) {
          affine.for %arg6 = #map0(%c0) to #map0(%2) {
            %8 = vector.splat %cst : vector<16xf32>
            memref.store %8, %7[%c0] : memref<1xvector<16xf32>>
            affine.for %arg7 = #map0(%c0) to #map0(%4) {
              %9 = affine.apply #map1(%5)
              affine.for %arg8 = #map0(%c0) to #map0(%9) step 2 {
                affine.for %arg9 = #map0(%c0) to #map0(%6) step 48 {
                  %10 = affine.apply #map2(%arg6, %arg8)
                  %11 = affine.apply #map0(%arg8)
                  %12 = affine.apply #map2(%arg5, %arg9)
                  %13 = affine.apply #map0(%arg9)
                  %cst_4 = arith.constant 0.000000e+00 : f32
                  %14 = vector.transfer_read %arg0[%arg3, %arg7, %10, %12], %cst_4 : memref<?x?x?x?xf32>, vector<16xf32>
                  %15 = affine.if #set(%11)[%5] -> vector<16xf32> {
                    %cst_10 = arith.constant 0.000000e+00 : f32
                    %47 = vector.transfer_read %arg0[%arg4, %arg7, %11, %13], %cst_10 : memref<?x?x?x?xf32>, vector<16xf32>
                    affine.yield %47 : vector<16xf32>
                  } else {
                    %47 = vector.splat %cst : vector<16xf32>
                    affine.yield %47 : vector<16xf32>
                  }
                  %16 = affine.apply #map3(%arg5, %arg9)
                  %17 = affine.apply #map4(%arg9)
                  %cst_5 = arith.constant 0.000000e+00 : f32
                  %18 = vector.transfer_read %arg0[%arg3, %arg7, %10, %16], %cst_5 : memref<?x?x?x?xf32>, vector<16xf32>
                  %19 = affine.if #set(%11)[%5] -> vector<16xf32> {
                    %cst_10 = arith.constant 0.000000e+00 : f32
                    %47 = vector.transfer_read %arg0[%arg4, %arg7, %11, %17], %cst_10 : memref<?x?x?x?xf32>, vector<16xf32>
                    affine.yield %47 : vector<16xf32>
                  } else {
                    %47 = vector.splat %cst : vector<16xf32>
                    affine.yield %47 : vector<16xf32>
                  }
                  %20 = affine.apply #map5(%arg5, %arg9)
                  %21 = affine.apply #map6(%arg9)
                  %cst_6 = arith.constant 0.000000e+00 : f32
                  %22 = vector.transfer_read %arg0[%arg3, %arg7, %10, %20], %cst_6 : memref<?x?x?x?xf32>, vector<16xf32>
                  %23 = affine.if #set(%11)[%5] -> vector<16xf32> {
                    %cst_10 = arith.constant 0.000000e+00 : f32
                    %47 = vector.transfer_read %arg0[%arg4, %arg7, %11, %21], %cst_10 : memref<?x?x?x?xf32>, vector<16xf32>
                    affine.yield %47 : vector<16xf32>
                  } else {
                    %47 = vector.splat %cst : vector<16xf32>
                    affine.yield %47 : vector<16xf32>
                  }
                  %24 = affine.apply #map7(%arg6, %arg8)
                  %25 = affine.apply #map8(%arg8)
                  %26 = affine.apply #map2(%arg5, %arg9)
                  %27 = affine.apply #map0(%arg9)
                  %cst_7 = arith.constant 0.000000e+00 : f32
                  %28 = vector.transfer_read %arg0[%arg3, %arg7, %24, %26], %cst_7 : memref<?x?x?x?xf32>, vector<16xf32>
                  %29 = affine.if #set(%25)[%5] -> vector<16xf32> {
                    %cst_10 = arith.constant 0.000000e+00 : f32
                    %47 = vector.transfer_read %arg0[%arg4, %arg7, %25, %27], %cst_10 : memref<?x?x?x?xf32>, vector<16xf32>
                    affine.yield %47 : vector<16xf32>
                  } else {
                    %47 = vector.splat %cst : vector<16xf32>
                    affine.yield %47 : vector<16xf32>
                  }
                  %30 = affine.apply #map3(%arg5, %arg9)
                  %31 = affine.apply #map4(%arg9)
                  %cst_8 = arith.constant 0.000000e+00 : f32
                  %32 = vector.transfer_read %arg0[%arg3, %arg7, %24, %30], %cst_8 : memref<?x?x?x?xf32>, vector<16xf32>
                  %33 = affine.if #set(%25)[%5] -> vector<16xf32> {
                    %cst_10 = arith.constant 0.000000e+00 : f32
                    %47 = vector.transfer_read %arg0[%arg4, %arg7, %25, %31], %cst_10 : memref<?x?x?x?xf32>, vector<16xf32>
                    affine.yield %47 : vector<16xf32>
                  } else {
                    %47 = vector.splat %cst : vector<16xf32>
                    affine.yield %47 : vector<16xf32>
                  }
                  %34 = affine.apply #map5(%arg5, %arg9)
                  %35 = affine.apply #map6(%arg9)
                  %cst_9 = arith.constant 0.000000e+00 : f32
                  %36 = vector.transfer_read %arg0[%arg3, %arg7, %24, %34], %cst_9 : memref<?x?x?x?xf32>, vector<16xf32>
                  %37 = affine.if #set(%25)[%5] -> vector<16xf32> {
                    %cst_10 = arith.constant 0.000000e+00 : f32
                    %47 = vector.transfer_read %arg0[%arg4, %arg7, %25, %35], %cst_10 : memref<?x?x?x?xf32>, vector<16xf32>
                    affine.yield %47 : vector<16xf32>
                  } else {
                    %47 = vector.splat %cst : vector<16xf32>
                    affine.yield %47 : vector<16xf32>
                  }
                  %38 = memref.load %7[%c0] : memref<1xvector<16xf32>>
                  %39 = vector.fma %14, %14, %38 : vector<16xf32>
                  %40 = vector.fma %18, %18, %39 : vector<16xf32>
                  %41 = vector.fma %22, %28, %40 : vector<16xf32>
                  %42 = vector.fma %28, %32, %41 : vector<16xf32>
                  memref.store %42, %7[%c0] : memref<1xvector<16xf32>>
                  %43 = memref.load %7[%c0] : memref<1xvector<16xf32>>
                  %44 = vector.reduction <add>, %43 : vector<16xf32> into f32
                  %45 = memref.load %arg0[%arg3, %arg4, %arg6, %arg5] : memref<?x?x?x?xf32>
                  %46 = arith.addf %45, %44 : f32
                  memref.store %46, %arg0[%arg3, %arg4, %arg6, %arg5] : memref<?x?x?x?xf32>
                }
              }
            }
          }
        }
      }
    }
    memref.dealloc %7 : memref<1xvector<16xf32>>
    return
  }
}
```

## Correctness
Correctness check now in `buddy-benchmark`, see [here](https://github.com/buddy-compiler/buddy-benchmark/pull/47) for more infomation.

### CODE REVIEW

**Review by hanliutong — CHANGES_REQUESTED** [review]
Hello, @WLFJ
I had a quick look at your patch and found some problems, please see the following:

1. [Makefile](https://github.com/buddy-compiler/buddy-mlir/tree/main/examples/MLIRLinalg/makefile) should be updated for the Conv2d and MatMul , taking MatMul as example:
```
linalg-matmul-optimize-lower:
	@${BUDDY_OPT} linalg-matmul.mlir ${MLIR_OPT_OPTIONS} \
		--matmul-optimize="vec-size=16 kernel-m=2 kernel-n=4" \
		-o ./log.mlir
linalg-matmul-optimize-translate:
...
linalg-matmul-optimize-run:
...
````
2. A new example about `conv_2d_nchw_fchw` should be added in MLIRLinalg
3. The result of MatMul seems wrong, please check your algorithm
```
root@1c1aaf706cfc:~/project/buddy-mlir/examples/MLIRLinalg# make linalg-matmul-optimize-run
Unranked Memref base@ = 0x56066eb7b8e0 rank = 2 offset = 0 sizes = [4, 4] strides = [4, 1] data = 
[[1,   1,   1,   1], 
 [1,   1,   1,   1], 
 [1,   1,   1,   1], 
 [1,   1,   1,   1]]
root@1c1aaf706cfc:~/project/buddy-mlir/examples/MLIRLinalg# make linalg-matmul-run
Unranked Memref base@ = 0x55c1b346d850 rank = 2 offset = 0 sizes = [4, 4] strides = [4, 1] data = 
[[5,   5,   5,   5], 
 [5,   5,   5,   5], 
 [5,   5,   5,   5], 
 [5,   5,   5,   5]]
```

**Comment by WLFJ** [comment]
Thanks for your review! I'll fix them tomorrow!

**Review by WLFJ — COMMENTED** [review]
Here's what I have done:
- [x] makefile update (both for MatMul and Conv2d)
- [x] fix MatMul and Conv running result error.
- [x] add Conv2d_NCHW_FCHW example.

File: lib/Conversion/ConvOptimization/ConvOptimize.cpp [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+//===----------------------------------------------------------------------===//
+//
+// This file implements the Conv optimize.
+//
+//===----------------------------------------------------------------------===//
+#include <iostream>
```
Comment (zhanghb97) [inline lib/Conversion/ConvOptimization/ConvOptimize.cpp:20]: This looks like an unnecessary include.

File: lib/Conversion/MatMulOptimization/MatMulOptimize.cpp [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+//===----------------------------------------------------------------------===//
+//
+// This file implements the matmul optimization.
+//
+//===----------------------------------------------------------------------===//
+#include <iostream>
```
Comment (zhanghb97) [inline lib/Conversion/MatMulOptimization/MatMulOptimize.cpp:20]: This looks like an unnecessary include.

File: lib/Conversion/MatMulOptimization/MatMulOptimize.cpp [hunk 3]
```diff
@@ -0,0 +1,269 @@
+//====- MatMulOptimize.cpp ----------------------------------------===//
```
Comment (zhanghb97) [inline lib/Conversion/MatMulOptimization/MatMulOptimize.cpp:1]: Wrong format: please align the lines of the header.

**Review by zhanghb97 — CHANGES_REQUESTED** [review]
Please use `clang-format` to format all the .cpp/.h files.

File: lib/Conversion/ConvOptimization/ConvOptimize.cpp (hunk 1 above)
Comment (WLFJ) [reply lib/Conversion/ConvOptimization/ConvOptimize.cpp]: Yes, I removed them.

File: lib/Conversion/MatMulOptimization/MatMulOptimize.cpp (hunk 2 above)
Comment (WLFJ) [reply lib/Conversion/MatMulOptimization/MatMulOptimize.cpp]: Yes, and also I checked and removed other unused include.

File: lib/Conversion/MatMulOptimization/MatMulOptimize.cpp (hunk 3 above)
Comment (WLFJ) [reply lib/Conversion/MatMulOptimization/MatMulOptimize.cpp]: Sorry for that, I check other cpp file header format and fixed it.

**Comment by WLFJ** [comment]
> Please use `clang-format` to format all the .cpp/.h files.

And I have already format using `clang-format`.
But the issue is in `ConvOptimize.cpp`. tools default 'Line width' is too short, so I set `clang-format` for a larger line width.

**Review by hanliutong — APPROVED** [review]

**Pull request merged**